test/distributed/fsdp/test_fsdp_use_orig_params.py

```diff
@@ -1048,10 +1049,122 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
+        Tests a basic ``no_sync()`` setup by comparing ``use_orig_params=True``
+        against ``use_orig_params=False``.
+        """
+        self.run_subtests(
+            {
+                "sharding_strategy": [
+                    ShardingStrategy.FULL_SHARD,
+                    ShardingStrategy.SHARD_GRAD_OP,
+                    ShardingStrategy.NO_SHARD,
+                ],
+            },
+            self._test_no_sync,
+        )
+
+    def _test_no_sync(self, sharding_strategy: ShardingStrategy):
+        model = nn.Linear(3, 3, bias=False, device="cuda")
+        fsdp_kwargs = {
+            "sharding_strategy": sharding_strategy,
+        }
+        baseline_model = FSDP(
+            copy.deepcopy(model), use_orig_params=False, **fsdp_kwargs
+        )
+        test_model = FSDP(model, use_orig_params=True, **fsdp_kwargs)
+        baseline_optim = torch.optim.AdamW(baseline_model.parameters(), foreach=True)
+        test_optim = torch.optim.AdamW(test_model.parameters(), foreach=True)
+
+        def _check_param_grad_parity(
+            _baseline_model: nn.Module,
+            _test_model: nn.Module,
+        ):
+            """
+            This assumes that the model is ``nn.Linear(3, 3, bias=False)``
+            (i.e. with a single weight parameter) to be able to directly
+            compare the baseline and test models. On rank 1, the baseline
+            includes 1 element of padding.
+            """
+            for baseline_param, test_param in zip(
+                _baseline_model.parameters(), _test_model.parameters()
+            ):
+                # Baseline is permitted to have padding
+                self.assertGreaterEqual(baseline_param.numel(), test_param.numel())
+                unpadded_param_numel = test_param.numel()
+                # For `NO_SHARD`, `use_orig_params=True` presents unflattened
+                # parameters, while `False` presents flattened ones
+                torch.testing.assert_close(
+                    baseline_param[:unpadded_param_numel], test_param.flatten()
+                )
+                # Gradient numel is different if right after `no_sync()` since
+                # the gradient is unsharded, while the parameter is sharded
+                unpadded_grad_numel = test_param.grad.numel()
+                # For `use_orig_params=False`, the unsharded gradient is
+                # flattened, while for `True`, it is unflattened
+                torch.testing.assert_close(
+                    baseline_param.grad[:unpadded_grad_numel].reshape(
+                        test_param.grad.shape
+                    ),
+                    test_param.grad,
+                )
+
+        inp = torch.randn((2, 3), device="cuda")
+        grad = torch.rand_like(inp)
+
+        # Compute some reference gradients using one forward/backward
+        baseline_out = baseline_model(inp)
+        test_out = test_model(inp)
+        torch.testing.assert_close(baseline_out, test_out)
```

Comment:
Is this more of a relaxed assert than `self.assertEqual`? When should we use one over the other?

Reply:
Good question. Let me investigate.